pubnub/crypto_core.py

```diff
@@ -12,3 +18,146 @@ def encrypt(self, key, msg):
+    CRYPTOR_VERSION: int = 1
+
+    @abstractmethod
+    def encrypt(self, data: bytes) -> CryptorPayload:
+        pass
+
+    @abstractmethod
+    def decrypt(self, payload: CryptorPayload, binary_mode: bool = False) -> bytes:
+        pass
+
+
+class PubNubLegacyCryptor(PubNubCryptor):
+    CRYPTOR_ID = '0000'
+    Initial16bytes = b'0123456789012345'
+
+    def __init__(self, cipher_key, use_random_iv=False, cipher_mode=AES.MODE_CBC, fallback_cipher_mode=None):
+        self.cipher_key = cipher_key
+        self.use_random_iv = use_random_iv
+        self.mode = cipher_mode
+        self.fallback_mode = fallback_cipher_mode
+
+    def encrypt(self, msg, *, key=None, use_random_iv=None):
+        key = key or self.cipher_key
+        use_random_iv = use_random_iv or self.use_random_iv
+
+        secret = self.get_secret(key)
+        initialization_vector = self.get_initialization_vector(use_random_iv)
+
+        cipher = AES.new(bytes(secret[0:32], 'utf-8'), self.mode, initialization_vector)
+        encrypted_message = cipher.encrypt(self.pad(msg))
+        msg_with_iv = self.append_random_iv(encrypted_message, use_random_iv, initialization_vector)
+        return CryptorPayload(data=msg_with_iv, cryptor_data=initialization_vector)
+
+    def decrypt(self, payload: CryptorPayload, key=None, use_random_iv=False, binary_mode: bool = False):
+        key = key or self.cipher_key
+        use_random_iv = use_random_iv or self.use_random_iv
+        secret = self.get_secret(key)
+        msg = payload['data']
+        initialization_vector, extracted_message = self.extract_random_iv(msg, use_random_iv)
+        cipher = AES.new(bytes(secret[0:32], "utf-8"), self.mode, initialization_vector)
+        if binary_mode:
+            return self.depad(cipher.decrypt(extracted_message), binary_mode)
+        try:
+            plain = self.depad((cipher.decrypt(extracted_message)).decode('utf-8'), binary_mode)
+        except UnicodeDecodeError as e:
+            if not self.fallback_mode:
+                raise e
+
+            cipher = AES.new(bytes(secret[0:32], "utf-8"), self.fallback_mode, initialization_vector)
+            plain = self.depad((cipher.decrypt(extracted_message)).decode('utf-8'), binary_mode)
+
+        try:
+            return json.loads(plain)
+        except Exception:
+            return plain
+
+    def append_random_iv(self, message, use_random_iv, initialization_vector):
+        if self.use_random_iv or use_random_iv:
+            return initialization_vector + message
+        else:
+            return message
+
+    def extract_random_iv(self, message, use_random_iv):
+        if use_random_iv:
+            return message[0:16], message[16:]
+        else:
+            return self.Initial16bytes, message
+
+    def get_initialization_vector(self, use_random_iv) -> bytes:
+        if self.use_random_iv or use_random_iv:
+            return bytes("{0:016}".format(random.randint(0, 9999999999999999)), 'utf-8')
```

Comment:
Please check `secrets` package for secure random source. It might hopefully also simplify this a little https://docs.python.org/3/library/secrets.html#secrets.token_bytes

Reply:
I won't change any logic in legacy cryptor. I want it to be as close to previous implementation as possible. In the `AesCbcCryptor` this method is updated though